### DISCUSSION TITLE
Add a `mapWithIndex` / `indexedMap` / `imap0` function

### DISCUSSION THREAD

**Discussion by ad-si** [typst/typst#3741] [General]
### Description

There should be a [`mapWithIndex`](https://hackage.haskell.org/package/containers-0.7/docs/Data-Sequence.html#v:mapWithIndex), or  [`indexedMap`](https://package.elm-lang.org/packages/elm/core/latest/List#indexedMap), or [`imap0`](https://noogle.dev/f/lib/imap0), … function.

Right now we only have the enumerate function:

```typ
#{
  let mapped = (1,2,3,4)
    .map(x => [value: #x \ ])
    .join()
  mapped
}

#{
  let enumed = for (index, value) in (1,2,3,4).enumerate() {
    [#index: #value \ ]
  }
  enumed
}

#{
  // TODO
  let mapped_idx = (1,2,3,4)
    .mapWithIndex((index, value) => 
      [#index: #value \ ]
    )
    .join()
  mapped_idx
}
```

### Use Case

Mapping with index in a pipeline. It's very unergonomic if you have to break the flow and use a for loop.

**Comment by Enivex** [comment]
What does this add over enumerate followed by map?

**Comment by ad-si** [comment]
I can't put it in pipelines like e.g:

```typ
#{
(("John", 175), ("anna", 168), ("bob", 183))
  .rev()
  .indexedMap((index, user) => {
    let (name, size) = user
    (index, upper(name), size / 100)
  })
  .flatten()
  .fold("", (a, e) => a + str(e))
}
```

**Comment by Enivex** [comment]
This works just fine
```typst
#{
(("John", 175), ("anna", 168), ("bob", 183))
  .rev()
  .enumerate()
  .map(((index, user)) => {
    let (name, size) = user
    (index, upper(name), size / 100)
  })
  .flatten()
  .fold("", (a, e) => a + str(e))
}
```
You can simplify it further though:
```typst
#{
(("John", 175), ("anna", 168), ("bob", 183))
  .rev()
  .enumerate()
  .map(((index, (name,size))) => {
    (index, upper(name), size / 100)
  })
  .flatten()
  .map(str).join()
}
```
Edit: Oops I overwrote my comment.

**Comment by ad-si** [comment]
I could have sworn that didn't work for whatever reason 😅🙈.
Makes it definitely less of a problem.

However, still feels quite inconsistent with all other pure fp languages, as they all have a `mapWithIndex` or similar.

**Reply by laurmaedje** [reply]
Probably you were missing one pair of parentheses. It's one destructured argument rather than two.

I think enumerate + map is more elegant than mapWithIndex though. It's also how Rust does it, to which Typst is closer than to most FPs.

**Reply by ad-si** [reply]
Ahh yes, that was the problem, thanks! I guess the error message could be better 😅
```txt
error: missing argument: row
    ┌─ typst/invoice-maker.typ:238:11
    │
238 │       .map((index, row) => {
    │            ^^^^^^^^^^^^
```

Ok sure, that's a good point.
I'm still wondering how one arrives at the Rust syntax coming from OCaml (which has the `mapi` function btw: https://v2.ocaml.org/api/List.html#VALmapi), but that's another discussion. 😛

**Reply by laurmaedje** [reply]
Yes, the error could be improved.
